File: app/src/main/java/com/cookplan/recipe/import_recipe/search_url/GoogleRecipeListRecyclerAdapter.java
package com.cookplan.recipe.import_recipe.search_url;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.cookplan.R;
import com.cookplan.models.network.GoogleRecipe;

import java.util.List;


public class GoogleRecipeListRecyclerAdapter extends RecyclerView.Adapter<GoogleRecipeListRecyclerAdapter.ViewHolder> {
    private List<GoogleRecipe> values;
    private GoogleRecipeListEventListener listener;
    private Context context;
    private boolean needToLoadNextPart;

    public GoogleRecipeListRecyclerAdapter(List<GoogleRecipe> items,
                                           GoogleRecipeListEventListener listener, Context context) {
        this.values = items;
        this.listener = listener;
        this.context = context;
        needToLoadNextPart = true;
    }

    @Override
    public GoogleRecipeListRecyclerAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.google_recipe_list_item_layout, parent, false);
        return new GoogleRecipeListRecyclerAdapter.ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(final GoogleRecipeListRecyclerAdapter.ViewHolder holder, int position) {
        GoogleRecipe item = values.get(position);
        holder.nameTextView.setText(item.getTitle());
        holder.descTextView.setText(item.getDesc());
        holder.urlTextView.setText(item.getFormattedUrl());
        if (item.getImageUrl() != null) {
            Glide.with(context)
                    .load(item.getImageUrl())
                    .placeholder(R.drawable.ic_default_recipe_image)
                    .centerCrop()
                    .into(holder.imageView);
        }

        holder.mainView.setTag(item);
        holder.mainView.setOnClickListener(v -> {
            GoogleRecipe localRecipe = (GoogleRecipe) v.getTag();
            if (listener != null && localRecipe != null) {
                listener.onItemClick(localRecipe.getUrl());
            }
        });
        if (position > values.size() * 2 / 3 && listener != null && needToLoadNextPart) {
            needToLoadNextPart = false;
            listener.onLoadNextData(values.size());
        }
    }

    @Override
    public int getItemCount() {
        return values.size();
    }

    public void addItems(List<GoogleRecipe> googleRecipes) {
        values.addAll(googleRecipes);
        notifyDataSetChanged();
        if (googleRecipes.size() == 0) {
            needToLoadNextPart = false;
        } else {
            needToLoadNextPart = true;
        }
    }

    public void clearItems() {
        values.clear();
        notifyDataSetChanged();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        public final View mainView;
        public final TextView nameTextView;
        public final TextView descTextView;
        public final TextView urlTextView;
        public final ImageView imageView;

        public ViewHolder(View view) {
            super(view);
            mainView = view.findViewById(R.id.card_view);
            nameTextView = (TextView) view.findViewById(R.id.name_textview);
            descTextView = (TextView) view.findViewById(R.id.desc_textview);
            urlTextView = (TextView) view.findViewById(R.id.link_textview);
            imageView = (ImageView) view.findViewById(R.id.recipe_image);
        }
    }

    public interface GoogleRecipeListEventListener {
        void onItemClick(String url);

        void onLoadNextData(int offset);
    }
}

File: app/src/main/java/com/cookplan/recipe/import_recipe/search_url/GoogleRecipeListRecyclerAdapter.kt
package com.cookplan.recipe.import_recipe.search_url

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.cookplan.R
import com.cookplan.models.network.GoogleRecipe
import com.cookplan.recipe.import_recipe.search_url.GoogleRecipeListRecyclerAdapter.GoogleRecipeViewHolder


class GoogleRecipeListRecyclerAdapter(private val values: MutableList<GoogleRecipe>,
                                      private val loadNextDatalistener: (Int) -> Unit,
                                      private val openUrlListener: (String) -> Unit) : RecyclerView.Adapter<GoogleRecipeViewHolder>() {
    private var needToLoadNextPart: Boolean = true

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): GoogleRecipeViewHolder =
            GoogleRecipeViewHolder(LayoutInflater.from(parent.context)
                    .inflate(R.layout.google_recipe_list_item_layout, parent, false))

    override fun onBindViewHolder(holderGoogleRecipe: GoogleRecipeViewHolder, position: Int) {
        val item = values[position]
        holderGoogleRecipe.nameTextView.text = item.title
        holderGoogleRecipe.descTextView.text = item.desc
        holderGoogleRecipe.urlTextView.text = item.formattedUrl
        if (item.imageUrl != null) {
            Glide.with(holderGoogleRecipe.imageView.context)
                    .load(item.imageUrl)
//                    .placeholder(R.drawable.ic_default_recipe_image)
//                    .centerCrop()
                    .into(holderGoogleRecipe.imageView)
        }

        with(holderGoogleRecipe.mainView) {
            tag = item
            setOnClickListener { if (item.url != null) openUrlListener(item.url) }
        }

        if (position > values.size * 2 / 3 && needToLoadNextPart) {
            needToLoadNextPart = false
            loadNextDatalistener(values.size)
        }
    }

    override fun getItemCount(): Int {
        return values.size
    }

    fun addItems(googleRecipes: List<GoogleRecipe>) {
        values.addAll(googleRecipes)
        notifyDataSetChanged()
        needToLoadNextPart = googleRecipes.size != 0
    }

    fun clearItems() {
        values.clear()
        notifyDataSetChanged()
    }

    inner class GoogleRecipeViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val mainView: View = view.findViewById(R.id.card_view)
        val nameTextView: TextView = view.findViewById<TextView>(R.id.name_textview)
        val descTextView: TextView = view.findViewById<TextView>(R.id.desc_textview)
        val urlTextView: TextView = view.findViewById<TextView>(R.id.link_textview)
        val imageView: ImageView = view.findViewById<ImageView>(R.id.recipe_image)
    }
}
